File: src/main/java/frc/robot/subsystems/SwerveSubsystem.java
package frc.robot.subsystems;

import static frc.robot.utils.RobotParameters.SwerveParameters.Thresholds.SHOULD_INVERT;

import com.ctre.phoenix6.hardware.Pigeon2;
import com.pathplanner.lib.auto.AutoBuilder;
import com.pathplanner.lib.config.PIDConstants;
import com.pathplanner.lib.controllers.PPHolonomicDriveController;
import edu.wpi.first.math.VecBuilder;
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.SwerveDriveKinematics;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.math.util.Units;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.DriverStation.Alliance;
import edu.wpi.first.wpilibj.smartdashboard.Field2d;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;
import frc.robot.utils.Dash;
import frc.robot.utils.PID;
import frc.robot.utils.RobotParameters.*;
import frc.robot.utils.RobotParameters.SwerveParameters.*;
import java.util.Optional;
import org.photonvision.EstimatedRobotPose;

/**
 * The SwerveSubsystem class represents the swerve drive subsystem of the robot. It handles the
 * control and state estimation of the swerve drive modules.
 */
public class SwerveSubsystem extends SubsystemBase {
  private final SwerveDrivePoseEstimator poseEstimator;
  private final Field2d field = new Field2d();
  private final Pigeon2 pidgey = new Pigeon2(MotorParameters.PIDGEY_ID);
  private final SwerveModuleState[] states = new SwerveModuleState[4];
  private final SwerveModule[] modules;
  private final Photonvision photonvision;
  private final PID pid;
  private double velocity = 0.0;

  /**
   * Constructs a new SwerveSubsystem.
   *
   * @param photonvision The Photonvision instance used for vision processing.
   */
  public SwerveSubsystem(Photonvision photonvision) {
    this.photonvision = photonvision;
    this.modules = initializeModules();
    this.pid = initializePID();
    this.pidgey.reset();
    this.poseEstimator = initializePoseEstimator();
    configureAutoBuilder();
  }

  /**
   * Initializes the swerve modules.
   *
   * @return An array of initialized SwerveModule objects.
   */
  private SwerveModule[] initializeModules() {
    return new SwerveModule[] {
      new SwerveModule(
          MotorParameters.FRONT_LEFT_DRIVE_ID,
          MotorParameters.FRONT_LEFT_STEER_ID,
          MotorParameters.FRONT_LEFT_CAN_CODER_ID,
          SwerveParameters.Thresholds.CANCODER_VAL9),
      new SwerveModule(
          MotorParameters.FRONT_RIGHT_DRIVE_ID,
          MotorParameters.FRONT_RIGHT_STEER_ID,
          MotorParameters.FRONT_RIGHT_CAN_CODER_ID,
          SwerveParameters.Thresholds.CANCODER_VAL10),
      new SwerveModule(
          MotorParameters.BACK_LEFT_DRIVE_ID,
          MotorParameters.BACK_LEFT_STEER_ID,
          MotorParameters.BACK_LEFT_CAN_CODER_ID,
          SwerveParameters.Thresholds.CANCODER_VAL11),
      new SwerveModule(
          MotorParameters.BACK_RIGHT_DRIVE_ID,
          MotorParameters.BACK_RIGHT_STEER_ID,
          MotorParameters.BACK_RIGHT_CAN_CODER_ID,
          SwerveParameters.Thresholds.CANCODER_VAL12)
    };
  }

  /**
   * Initializes the PID controller.
   *
   * @return A new PID object with values from the SmartDashboard.
   */
  private PID initializePID() {
    return new PID(
        SmartDashboard.getNumber("AUTO: P", PIDParameters.DRIVE_PID_AUTO.getP()),
        SmartDashboard.getNumber("AUTO: I", PIDParameters.DRIVE_PID_AUTO.getI()),
        SmartDashboard.getNumber("AUTO: D", PIDParameters.DRIVE_PID_AUTO.getD()));
  }

  /**
   * Initializes the SwerveDrivePoseEstimator.
   *
   * @return A new SwerveDrivePoseEstimator object.
   */
  private SwerveDrivePoseEstimator initializePoseEstimator() {
    return new SwerveDrivePoseEstimator(
        SwerveParameters.PhysicalParameters.kinematics,
        Rotation2d.fromDegrees(getHeading()),
        getModulePositions(),
        new Pose2d(0.0, 0.0, Rotation2d.fromDegrees(0.0)),
        VecBuilder.fill(0.05, 0.05, Units.degreesToRadians(5.0)),
        VecBuilder.fill(0.5, 0.5, Units.degreesToRadians(30.0)));
  }

  /** Configures the AutoBuilder for autonomous driving. */
  private void configureAutoBuilder() {
    AutoBuilder.configure(
        this::getPose,
        this::newPose,
        this::getAutoSpeeds,
        this::chassisSpeedsDrive,
        new PPHolonomicDriveController(
            new PIDConstants(5.0, 0.0, 0.0), new PIDConstants(5.0, 0.0, 0.0)),
        PIDParameters.config,
        () -> {
          Optional<Alliance> alliance = DriverStation.getAlliance();
          if (alliance.isEmpty()) {
            return false;
          }
          if (SHOULD_INVERT) {
            return alliance.get() == Alliance.Red;
          } else {
            return alliance.get() != Alliance.Blue;
          }
        },
        this);
  }

  /**
   * This method is called periodically by the scheduler. It updates the pose estimator and
   * dashboard values.
   */
  @Override
  public void periodic() {

    /**
     * This method checks whether the bot is in Teleop, and adds it to poseEstimator based on VISION
     */
    if (DriverStation.isTeleop()) {
      EstimatedRobotPose estimatedPose =
          photonvision.getEstimatedGlobalPose(poseEstimator.getEstimatedPosition());
      if (estimatedPose != null) {
        double timestamp = estimatedPose.timestampSeconds;
        Pose2d visionMeasurement2d = estimatedPose.estimatedPose.toPose2d();
        poseEstimator.addVisionMeasurement(visionMeasurement2d, timestamp);
        SwerveParameters.Thresholds.currentPose = poseEstimator.getEstimatedPosition();
      }
    }

    /**
     * Updates the robot position based on movement and rotation from the pidgey and encoders
     */
    poseEstimator.update(getPidgeyRotation(), getModulePositions());

    field.setRobotPose(poseEstimator.getEstimatedPosition());

    Dash.dash(
        Dash.pairOf("Pitch", pidgey.getPitch().getValueAsDouble()),
        Dash.pairOf("Heading", -pidgey.getYaw().getValueAsDouble()),
        Dash.pairOf("Yaw", pidgey.getYaw().getValueAsDouble()),
        Dash.pairOf("Roll", pidgey.getRoll().getValueAsDouble()),
        Dash.pairOf("Robot Pose", field));
  }

  /**
   * Sets the drive speeds for the swerve modules.
   *
   * @param forwardSpeed The forward speed.
   * @param leftSpeed The left speed.
   * @param turnSpeed The turn speed.
   * @param isFieldOriented Whether the drive is field-oriented.
   */
  public void setDriveSpeeds(
      double forwardSpeed, double leftSpeed, double turnSpeed, boolean isFieldOriented) {
    Dash.dash(
        Dash.pairOf("Forward speed", forwardSpeed),
        Dash.pairOf("Left speed", leftSpeed),
        Dash.pairOf("Pidgey Heading", getHeading()));

    ChassisSpeeds speeds = new ChassisSpeeds(forwardSpeed, leftSpeed, turnSpeed);
    if (!isFieldOriented) {
      speeds.toRobotRelativeSpeeds(getPidgeyRotation());
    }

    SwerveModuleState[] states2 =
        SwerveParameters.PhysicalParameters.kinematics.toSwerveModuleStates(speeds);
    SwerveDriveKinematics.desaturateWheelSpeeds(states2, MotorParameters.MAX_SPEED);
    setModuleStates(states2);
  }

  /**
   * Gets the rotation of the Pigeon2 IMU.
   *
   * @return The rotation of the Pigeon2 IMU.
   */
  public Rotation2d getPidgeyRotation() {
    return pidgey.getRotation2d();
  }

  /**
   * Gets the heading of the robot.
   *
   * @return The heading of the robot.
   */
  public double getHeading() {
    return -pidgey.getYaw().getValueAsDouble();
  }

  /**
   * Gets the yaw of the Pigeon2 IMU.
   *
   * @return The yaw of the Pigeon2 IMU.
   */
  public double getPidgeyYaw() {
    return pidgey.getYaw().getValueAsDouble();
  }

  /** Sets the initial heading of the robot based on the alliance color. */
  public void setInitialHeading() {
    Optional<Alliance> alliance = DriverStation.getAlliance();
    if (alliance.isEmpty() || alliance.get() == Alliance.Red) {
      pidgey.setYaw(27.4);
    } else {
      pidgey.setYaw(-27.4);
    }
  }

  /** Resets the Pigeon2 IMU. */
  public void resetPidgey() {
    pidgey.reset();
  }

  /**
   * Gets the current pose of the robot.
   *
   * @return The current pose of the robot.
   */
  public Pose2d getPose() {
    return poseEstimator.getEstimatedPosition();
  }

  /** Resets the pose of the robot to zero. */
  public void zeroPose() {
    poseEstimator.resetPosition(
        Rotation2d.fromDegrees(getHeading()),
        getModulePositions(),
        new Pose2d(0.0, 0.0, Rotation2d.fromDegrees(0.0)));
  }

  /**
   * Sets a new pose for the robot.
   *
   * @param pose The new pose.
   */
  public void newPose(Pose2d pose) {
    poseEstimator.resetPosition(Rotation2d.fromDegrees(getHeading()), getModulePositions(), pose);
  }

  /**
   * Gets the chassis speeds for autonomous driving.
   *
   * @return The chassis speeds for autonomous driving.
   */
  public ChassisSpeeds getAutoSpeeds() {
    SwerveDriveKinematics k = SwerveParameters.PhysicalParameters.kinematics;
    return k.toChassisSpeeds(getModuleStates());
  }

  /**
   * Gets the rotation of the Pigeon2 IMU for PID control.
   *
   * @return The rotation of the Pigeon2 IMU for PID control.
   */
  public Rotation2d getRotationPidggy() {
    return Rotation2d.fromDegrees(-pidgey.getRotation2d().getDegrees());
  }

  /**
   * Drives the robot using chassis speeds.
   *
   * @param chassisSpeeds The chassis speeds.
   */
  public void chassisSpeedsDrive(ChassisSpeeds chassisSpeeds) {
    SwerveModuleState[] newStates =
        SwerveParameters.PhysicalParameters.kinematics.toSwerveModuleStates(chassisSpeeds);
    setModuleStates(newStates);
  }

  /**
   * Gets the states of the swerve modules.
   *
   * @return The states of the swerve modules.
   */
  public SwerveModuleState[] getModuleStates() {
    SwerveModuleState[] moduleStates = new SwerveModuleState[modules.length];
    for (int i = 0; i < modules.length; i++) {
      moduleStates[i] = modules[i].getState();
    }
    return moduleStates;
  }

  /**
   * Sets the states of the swerve modules.
   *
   * @param states The states of the swerve modules.
   */
  public void setModuleStates(SwerveModuleState[] states) {
    for (int i = 0; i < states.length; i++) {
      modules[i].setState(states[i]);
    }
  }

  /**
   * Gets the positions of the swerve modules.
   *
   * @return The positions of the swerve modules.
   */
  public SwerveModulePosition[] getModulePositions() {
    SwerveModulePosition[] positions = new SwerveModulePosition[states.length];
    for (int i = 0; i < positions.length; i++) {
      positions[i] = modules[i].getPosition();
    }
    return positions;
  }

  /** Stops all swerve modules. */
  public void stop() {
    for (SwerveModule module : modules) {
      module.stop();
    }
  }

  /** Sets the PID constants for autonomous driving. */
  public void setAutoPID() {
    for (SwerveModule module : modules) {
      module.setAUTOPID();
    }
  }

  /** Sets the PID constants for teleoperated driving. */
  public void setTelePID() {
    for (SwerveModule module : modules) {
      module.setTELEPID();
    }
  }

  /** Resets the drive positions of all swerve modules. */
  public void resetDrive() {
    for (SwerveModule module : modules) {
      module.resetDrivePosition();
    }
  }

  /** Sets custom PID constants for the drive. */
  public void setCustomDrivePID() {
    Dash.dashPID("Drive", pid, PIDParameters.DRIVE_PID_V_AUTO, v -> velocity = v);
    for (SwerveModule module : modules) {
      module.setDrivePID(pid, velocity);
    }
  }
}

File: src/main/java/frc/robot/subsystems/SwerveSubsystem.kt
@file:Suppress("MemberVisibilityCanBePrivate", "unused")

package frc.robot.subsystems

import com.ctre.phoenix6.hardware.Pigeon2
import com.pathplanner.lib.auto.AutoBuilder
import com.pathplanner.lib.config.PIDConstants
import com.pathplanner.lib.controllers.PPHolonomicDriveController
import edu.wpi.first.math.VecBuilder
import edu.wpi.first.math.estimator.SwerveDrivePoseEstimator
import edu.wpi.first.math.geometry.Pose2d
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.math.kinematics.ChassisSpeeds
import edu.wpi.first.math.kinematics.SwerveDriveKinematics
import edu.wpi.first.math.kinematics.SwerveModulePosition
import edu.wpi.first.math.kinematics.SwerveModuleState
import edu.wpi.first.math.util.Units
import edu.wpi.first.wpilibj.DriverStation
import edu.wpi.first.wpilibj.DriverStation.Alliance
import edu.wpi.first.wpilibj.smartdashboard.Field2d
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import edu.wpi.first.wpilibj2.command.SubsystemBase
import frc.robot.utils.PID
import frc.robot.utils.RobotParameters.MotorParameters
import frc.robot.utils.RobotParameters.SwerveParameters
import frc.robot.utils.RobotParameters.SwerveParameters.PIDParameters
import frc.robot.utils.RobotParameters.SwerveParameters.Thresholds.SHOULD_INVERT
import frc.robot.utils.dash
import frc.robot.utils.dashPID

/** The [SwerveSubsystem] class includes all the motors to drive the robot. */
class SwerveSubsystem(photonvision: Photonvision) : SubsystemBase() {
  private val poseEstimator: SwerveDrivePoseEstimator
  private val field = Field2d()

  private val pidgey = Pigeon2(MotorParameters.PIDGEY_ID)
  private val states: Array<SwerveModuleState?>
  private val modules =
    arrayOf(
      SwerveModule(
        MotorParameters.FRONT_LEFT_DRIVE_ID,
        MotorParameters.FRONT_LEFT_STEER_ID,
        MotorParameters.FRONT_LEFT_CAN_CODER_ID,
        SwerveParameters.Thresholds.CANCODER_VAL9,
      ),
      SwerveModule(
        MotorParameters.FRONT_RIGHT_DRIVE_ID,
        MotorParameters.FRONT_RIGHT_STEER_ID,
        MotorParameters.FRONT_RIGHT_CAN_CODER_ID,
        SwerveParameters.Thresholds.CANCODER_VAL10,
      ),
      SwerveModule(
        MotorParameters.BACK_LEFT_DRIVE_ID,
        MotorParameters.BACK_LEFT_STEER_ID,
        MotorParameters.BACK_LEFT_CAN_CODER_ID,
        SwerveParameters.Thresholds.CANCODER_VAL11,
      ),
      SwerveModule(
        MotorParameters.BACK_RIGHT_DRIVE_ID,
        MotorParameters.BACK_RIGHT_STEER_ID,
        MotorParameters.BACK_RIGHT_CAN_CODER_ID,
        SwerveParameters.Thresholds.CANCODER_VAL12,
      ),
    )

  private val photonvision: Photonvision

  private val pid =
    PID(
      SmartDashboard.getNumber("AUTO: P", PIDParameters.DRIVE_PID_AUTO.p),
      SmartDashboard.getNumber("AUTO: I", PIDParameters.DRIVE_PID_AUTO.i),
      SmartDashboard.getNumber("AUTO: D", PIDParameters.DRIVE_PID_AUTO.d),
    )
  private var velocity = 0.0

  init {
    pidgey.reset()
    states = arrayOfNulls(4)
    this.photonvision = photonvision

    poseEstimator =
      SwerveDrivePoseEstimator(
        SwerveParameters.PhysicalParameters.kinematics,
        Rotation2d.fromDegrees(heading),
        modulePositions,
        Pose2d(0.0, 0.0, Rotation2d.fromDegrees(0.0)),
        VecBuilder.fill(0.05, 0.05, Units.degreesToRadians(5.0)),
        VecBuilder.fill(0.5, 0.5, Units.degreesToRadians(30.0)),
      )

    AutoBuilder.configure(
      { this.pose }, // Robot pose supplier
      { pose: Pose2d? ->
        this.newPose(pose)
      }, // Method to reset odometry (will be called if your auto has a starting pose)
      { this.autoSpeeds }, // ChassisSpeeds supplier. MUST BE ROBOT RELATIVE
      { chassisSpeeds: ChassisSpeeds? ->
        this.chassisSpeedsDrive(chassisSpeeds)
      }, // Method that will drive the robot given ROBOT RELATIVE
      PPHolonomicDriveController( // PPHolonomicController is the built-in path following
        // controller for holonomic drive trains
        PIDConstants(5.0, 0.0, 0.0), // Translation PID constants
        PIDConstants(5.0, 0.0, 0.0), // Rotation PID constants
      ),
      PIDParameters.config,
      {
        val alliance = DriverStation.getAlliance()
        alliance
          .map {
            if (SHOULD_INVERT) {
              it == Alliance.Red
            } else {
              it != Alliance.Blue
            }
          }
          .orElse(false)
      },
      this, // Reference to this subsystem to set requirements
    )
  }

  /**
   * This method will be called once per scheduler run.
   * It updates the robot's pose estimation and logs relevant data to the dashboard.
   */
  override fun periodic() {
    if (DriverStation.isTeleop()) {
      photonvision.getEstimatedGlobalPose(poseEstimator.estimatedPosition)?.apply {
        val timestamp = timestampSeconds
        val estimatedPose = estimatedPose
        val visionMeasurement2d = estimatedPose.toPose2d()
        poseEstimator.addVisionMeasurement(visionMeasurement2d, timestamp)
        SwerveParameters.Thresholds.currentPose = poseEstimator.estimatedPosition
      }
    }

    poseEstimator.update(pidgeyRotation, modulePositions)

    field.robotPose = poseEstimator.estimatedPosition

    dash(
      "Pitch" to pidgey.pitch.valueAsDouble,
      "Heading" to -pidgey.yaw.valueAsDouble,
      "Yaw" to pidgey.yaw.valueAsDouble,
      "Roll" to pidgey.roll.valueAsDouble,
      "Robot Pose" to field,
    )
  }

  /**
   * Sets the drive speeds for the robot.
   *
   * @param forwardSpeed The forward speed.
   * @param leftSpeed The left speed.
   * @param turnSpeed The turn speed.
   * @param isFieldOriented Whether the drive is field-oriented.
   */
  fun setDriveSpeeds(
    forwardSpeed: Double,
    leftSpeed: Double,
    turnSpeed: Double,
    isFieldOriented: Boolean,
  ) {
    dash("Forward speed" to forwardSpeed, "Left speed" to leftSpeed, "Pidgey Heading" to heading)

    val speeds = ChassisSpeeds(forwardSpeed, leftSpeed, turnSpeed).apply {
      if (!isFieldOriented) toRobotRelativeSpeeds(pidgeyRotation)
    }

    val states2 = SwerveParameters.PhysicalParameters.kinematics.toSwerveModuleStates(speeds)
    SwerveDriveKinematics.desaturateWheelSpeeds(states2, MotorParameters.MAX_SPEED)
    moduleStates = states2
  }

  val pidgeyRotation: Rotation2d
    /**
     * Gets the pidgey rotation.
     *
     * @return [Rotation2d] The current rotation of the [pidgey].
     */
    get() = pidgey.rotation2d

  val heading: Double
    /**
     * Gets the pidgey angle.
     *
     * @return [Double] The current angle of the [pidgey].
     */
    get() = -pidgey.yaw.valueAsDouble

  val pidgeyYaw: Double
    /**
     * Gets the pidgey yaw.
     *
     * @return [Double] The current yaw of the [pidgey].
     */
    get() = pidgey.yaw.valueAsDouble

  fun setHeading() {
    val alliance = DriverStation.getAlliance()
    if (alliance.get() == Alliance.Red) {
      pidgey.setYaw(27.4)
    } else {
      pidgey.setYaw(-27.4)
    }
  }

  /** Resets the pidgey(gyro) heading to 0. */
  fun resetPidgey() {
    pidgey.reset()
  }

  val pose: Pose2d
    /**
     * Gets the current pose of the robot.
     *
     * @return [Pose2d] The current pose of the robot.
     */
    get() = poseEstimator.estimatedPosition

  /** Zeros the pose of the robot. */
  fun zeroPose() {
    poseEstimator.resetPosition(
      Rotation2d.fromDegrees(heading),
      modulePositions,
      Pose2d(0.0, 0.0, Rotation2d.fromDegrees(0.0)),
    )
  }

  /**
   * Resets the pose of the robot.
   *
   * @param pose The new pose to set.
   */
  fun newPose(pose: Pose2d?) {
    poseEstimator.resetPosition(Rotation2d.fromDegrees(heading), modulePositions, pose)
  }

  val autoSpeeds: ChassisSpeeds
    /**
     * Gets the chassis speeds for autonomous driving.
     *
     * @return [ChassisSpeeds] The current chassis speeds.
     */
    get() {
      val k = SwerveParameters.PhysicalParameters.kinematics
      // Necessary because a kotlin array is being passed into a java varargs parameter
      val chassisSpeeds =
        k::class.java.getMethod("toChassisSpeeds", Array<SwerveModuleState>::class.java)
      return chassisSpeeds.invoke(k, moduleStates) as ChassisSpeeds
    }

  val rotationPidggy: Rotation2d
    /**
     * Gets the rotation of the pidgey for PID control.
     *
     * @return [Rotation2d] The rotation of the pidgey for PID control.
     */
    get() = Rotation2d.fromDegrees(-pidgey.rotation2d.degrees)

  /**
   * Drives the robot using the given chassis speeds.
   *
   * @param chassisSpeeds The desired chassis speeds.
   */
  fun chassisSpeedsDrive(chassisSpeeds: ChassisSpeeds?) {
    // ChassisSpeeds speeds =
    //     ChassisSpeeds.fromRobotRelativeSpeeds(chassisSpeeds, getRotationPidggy());
    // SwerveModuleState[] newStates = SwerveGlobalValues.kinematics.toSwerveModuleStates(speeds);
    // SwerveDriveKinematics.desaturateWheelSpeeds(newStates, MotorGlobalValues.MAX_SPEED);

    val newStates =
      SwerveParameters.PhysicalParameters.kinematics.toSwerveModuleStates(chassisSpeeds)
    moduleStates = newStates
  }

  private var moduleStates: Array<SwerveModuleState>
    /**
     * Gets the current states of the swerve modules.
     *
     * @return SwerveModuleState[] The current states of the swerve modules.
     */
    get() {
      val moduleStates = emptyArray<SwerveModuleState>()
      for (i in modules.indices) {
        moduleStates[i] = modules[i].state
      }
      return moduleStates
    }
    /**
     * Sets the desired module states.
     *
     * @param states SwerveModuleState[] The desired states for the swerve modules.
     */
    set(states) {
      for (i in states.indices) {
        modules[i].state = states[i]
      }
    }

  val modulePositions: Array<SwerveModulePosition?>
    /**
     * Gets the current positions of the swerve modules.
     *
     * @return SwerveModulePosition[] The current positions of the swerve modules.
     */
    get() {
      val positions = arrayOfNulls<SwerveModulePosition>(states.size)

      for (i in positions.indices) {
        positions[i] = modules[i].position
      }

      return positions
    }

  /** Stops the robot. */
  fun stop() = modules.forEach { it.stop() }

  fun setAutoPID() = modules.forEach { it.setAUTOPID() }

  fun setTelePID() = modules.forEach { it.setTELEPID() }

  fun resetDrive() = modules.forEach { it.resetDrivePosition() }

  fun setCustomDrivePID() {
    dashPID("Drive", pid, PIDParameters.DRIVE_PID_V_AUTO) { velocity = it }
    modules.forEach { it.setAUTOPID(pid, velocity) }
  }
}
